Improve operation workflow concurrency model (#2527)
thin-edge/thin-edge.io: docs/src/references/agent/operation-workflow.md

on_error = "failed_restart"
```

#### Awaiting the agent to restart

> When the expected outcome of a script triggers is to restart the device or the agent,

Author: fixed

this script cannot be monitored end-to-end by the agent which will die executing the script.
The success or failure of the script can only be detected when the agent resumes.

> This is done using a combination of the `await-agent-restart` action with a `background_script`.

Author: Added

1. The script is declared as a `background_script` with an `on_exec` handler.

>    This handler tells the agent to which state to move on before executing the script.

Author: Fixed

   The agent persists on disk this next state before launching the script which outcome can be a reboot.

> 2. To action attached to this restarting state is to `await-agent-restart`
>    with two handlers for the successful and failed cases.

Reviewer: ```suggestion
2. The action attached to this next state is to `await-agent-restart`
   with two handlers for the successful and timeout cases.
```

Author: Fixed

> 3. On a successful reboot,the agent resumes this persisted state when restarting and simply moves to the successful case.

Reviewer: ```suggestion
3. On a successful reboot, the agent resumes from this persisted state awaiting restart and simply moves to the successful case.
```

Author: fixed

> 4. If for some reason, no restart happens before the given delay, the agent moves the `on_timeout` operation state. 

Reviewer: ```suggestion
4. If for some reason, no restart happens within the given timeout window, the agent moves to the `on_timeout` operation state. 
```

Author: changed
